### PR TITLE
hyprland/workspaces improvements

### PR DESCRIPTION
**Pull request by MightyPlaza**
Some improvements to the hyprland/workspaces modules

>adds workspace clicking support (except for the non named special workspace which seems to be an issue on hyprland's side)

>adds "all-ouptuts" and "show-special" to config

>listens to "focusedmon" and "moveworkspace" events

>actually implements "{name}" format which was  present on the wiki but wasn't implemented (opposite of  "{id}")

>fixes issues with previous event implementation on named workspaces

>adds more variables to the workspaces

>sorts workspaces in the order normal -> named -> special -> named special, instead of id

didn't find any other issues/bugs 

will also ping @Anakael @zjeffer in case I accidentally broke something

### CODE REVIEW

**Comment by zjeffer** [comment]
Great, thanks for this! I'll do some tests with this tomorrow with multiple monitors.

I just compiled and ran it on a single monitor, and I had a segfault after switching to the fourth workspace, when it creates the new workspace button. I can't seem recreate it again, so I'll do some more testing.

File: include/modules/hyprland/workspaces.hpp [hunk 1]
```diff
@@ -9,27 +9,30 @@
...
   Gtk::Button& button() { return button_; };
 
+  int id() const { return id_; };
+  std::string name() const { return name_; };
+  std::string monitor() const { return monitor_; };
+  int active() const { return active_; };
+  bool is_special() const { return is_special_; };
+
+  auto handle_clicked(GdkEventButton* bt) -> bool;
+  void set_active(bool value) { active_ = value; };
```
Comment (Anakael) [inline include/modules/hyprland/workspaces.hpp:25]: I can suggest put `value = true` here.

File: src/modules/hyprland/workspaces.cpp [hunk 2]
```diff
@@ -56,55 +68,73 @@ auto Workspaces::update() -> void {
...
   } else if (eventName == "createworkspace") {
-    int new_workspace_id;
-    std::from_chars(payload.data(), payload.data() + payload.size(), new_workspace_id);
-    workspaces_to_create_.push_back(new_workspace_id);
+    const Json::Value workspaces_json = gIPC->getSocket1JsonReply("workspaces");
+    for (auto &workspace_json : workspaces_json) {
+      if (workspace_json["name"].asString() == payload &&
+          (all_outputs() || bar_.output->name == workspace_json["monitor"].asString()) &&
+          (workspace_json["name"].asString().find("special:") != 0 || show_special()))
+        create_workspace(workspace_json);
```
Comment (Anakael) [inline src/modules/hyprland/workspaces.cpp:99]: You can't create gtk entities here. Only in `update` method.

**Review by Anakael — CHANGES_REQUESTED** [review]

File: src/modules/hyprland/workspaces.cpp (hunk 2 above)
Comment (zjeffer) [reply src/modules/hyprland/workspaces.cpp:99]: This is probably the reason for the segfault

File: src/modules/hyprland/workspaces.cpp [hunk 3]
```diff
@@ -193,7 +249,20 @@ std::string &Workspace::select_icon(std::map<std::string, std::string> &icons_ma
   if (default_icon_it != icons_map.end()) {
     return default_icon_it->second;
   }
-
   return icons_map[""];
 }
+
+auto Workspace::handle_clicked(GdkEventButton *bt) -> bool {
+  if (id() > 0) {  // normal
+    system(("hyprctl dispatch workspace " + std::to_string(id()) + " &> /dev/null").c_str());
```
Comment (zjeffer) [inline src/modules/hyprland/workspaces.cpp:257]: Instead of using system(), we should do this through hyprland's IPC.

Comment (zjeffer) [reply src/modules/hyprland/workspaces.cpp:257]: I made a PR with this change here: https://github.com/MightyPlaza/Waybar/pull/1

File: src/modules/hyprland/workspaces.cpp (hunk 2 above)
Comment (Anakael) [reply src/modules/hyprland/workspaces.cpp:99]: Yes, exactly.

File: man/waybar-hyprland-workspaces.5.scd [hunk 4]
```diff
@@ -48,7 +50,8 @@ Additional to workspace name matching, the following *format-icons* can be set.
 		"active": "",
 		"default": ""
 	},
-	"sort-by-number": true
+	"all-outputs": false,
+        "show-special": false
```
Comment (Anakael) [inline man/waybar-hyprland-workspaces.5.scd:54]: Can you please explain in doc what it is?

File: include/modules/hyprland/workspaces.hpp [hunk 5]
```diff
@@ -9,27 +9,30 @@
...
+
   void update(const std::string& format, const std::string& icon);
 
  private:
   int id_;
+  std::string name_;
+  std::string monitor_;
+  int windows_;
```
Comment (Anakael) [inline include/modules/hyprland/workspaces.hpp:33]: What it is?

File: include/modules/hyprland/workspaces.hpp [hunk 6]
```diff
@@ -43,19 +46,27 @@ class Workspaces : public AModule, public EventHandler {
   void update() override;
   void init();
 
+  auto all_outputs() const -> bool { return all_outputs_; }
+  auto show_special() const -> bool { return show_special_; }
```
Comment (Anakael) [inline include/modules/hyprland/workspaces.hpp:50]: I think it's better to put it in Workspaces. Not individual workspace. Because it's value is same for all workspaces.

File: include/modules/hyprland/workspaces.hpp [hunk 7]
```diff
@@ -43,19 +46,27 @@ class Workspaces : public AModule, public EventHandler {
...
+  void remove_workspace(std::string name);
+
+  bool all_outputs_ = false;
+  bool show_special_ = false;
 
   std::string format_;
   std::map<std::string, std::string> icons_map_;
   bool with_icon_;
-  int active_workspace_id;
+  std::string active_workspace_name;
```
Comment (Anakael) [inline include/modules/hyprland/workspaces.hpp:66]: What's the reason for this change and for workspaces to remove?
Comparing integers is faster than strings.

File: include/modules/hyprland/workspaces.hpp [hunk 8]
```diff
@@ -43,19 +46,27 @@ class Workspaces : public AModule, public EventHandler {
...
+
+  auto get_bar_output() const -> std::string { return bar_.output->name; }
+
  private:
   void onEvent(const std::string&) override;
   void sort_workspaces();
-  void create_workspace(int id);
-  void remove_workspace(int id);
+  void create_workspace(const Json::Value& value);
+  void remove_workspace(std::string name);
```
Comment (Anakael) [inline include/modules/hyprland/workspaces.hpp:58]: Why change id to name here?

File: include/modules/hyprland/workspaces.hpp [hunk 9]
```diff
@@ -43,19 +46,27 @@ class Workspaces : public AModule, public EventHandler {
...
+  auto show_special() const -> bool { return show_special_; }
+
+  auto get_bar_output() const -> std::string { return bar_.output->name; }
+
  private:
   void onEvent(const std::string&) override;
   void sort_workspaces();
-  void create_workspace(int id);
-  void remove_workspace(int id);
+  void create_workspace(const Json::Value& value);
```
Comment (Anakael) [inline include/modules/hyprland/workspaces.hpp:57]: I would prefer leave Dto struct. Json value hints nothing about what's inside it, but Dto does.
And the reason why id was here -- is because it's clearly to see what's going on inside `onEvent` methond and `update` method. Push Json value through methods -- it' a mess. You can compare current solution with sway solution.

File: src/modules/hyprland/workspaces.cpp [hunk 10]
```diff
@@ -56,55 +68,73 @@ auto Workspaces::update() -> void {
     create_workspace(workspace_to_create);
   }
 
-  workspaces_to_create_.clear();
-
-  for (std::unique_ptr<Workspace> &workspace : workspaces_) {
-    workspace->set_active(workspace->id() == active_workspace_id);
-
+  for (auto &workspace : workspaces_) {
+    workspace->set_active(workspace->name() == active_workspace_name);
```
Comment (Anakael) [inline src/modules/hyprland/workspaces.cpp:72]: It's better to leave it with integers comparision.

File: src/modules/hyprland/workspaces.cpp [hunk 11]
```diff
@@ -56,55 +68,73 @@ auto Workspaces::update() -> void {
...
+
   } else if (eventName == "createworkspace") {
-    int new_workspace_id;
-    std::from_chars(payload.data(), payload.data() + payload.size(), new_workspace_id);
-    workspaces_to_create_.push_back(new_workspace_id);
+    const Json::Value workspaces_json = gIPC->getSocket1JsonReply("workspaces");
+    for (auto &workspace_json : workspaces_json) {
+      if (workspace_json["name"].asString() == payload &&
+          (all_outputs() || bar_.output->name == workspace_json["monitor"].asString()) &&
+          (workspace_json["name"].asString().find("special:") != 0 || show_special()))
```
Comment (Anakael) [inline src/modules/hyprland/workspaces.cpp:98]: Can you please explain what special workspace is?
Didn't found any info in Hyprland wiki.

File: include/modules/hyprland/workspaces.hpp [hunk 12]
```diff
@@ -9,27 +9,30 @@
...
+
+  auto handle_clicked(GdkEventButton* bt) -> bool;
+  void set_active(bool value) { active_ = value; };
+
   void update(const std::string& format, const std::string& icon);
 
  private:
   int id_;
+  std::string name_;
+  std::string monitor_;
```
Comment (Anakael) [inline include/modules/hyprland/workspaces.hpp:32]: I can suggest name it `output`, as in wayland ecosystem.

File: src/modules/hyprland/workspaces.cpp [hunk 13]
```diff
@@ -56,55 +68,73 @@ auto Workspaces::update() -> void {
...
+        create_workspace(workspace_json);
+    }
+
+  } else if (eventName == "focusedmon") {
+    active_workspace_name = payload.substr(payload.find(",") + 1);
+
+  } else if (eventName == "moveworkspace") {
+    std::string workspace = payload.substr(0, payload.find(","));
+    std::string new_monitor = payload.substr(payload.find(",") + 1);
+    if (all_outputs()) {
```
Comment (Anakael) [inline src/modules/hyprland/workspaces.cpp:108]: You can revert `if` statement and remove one level of indention.

File: src/modules/hyprland/workspaces.cpp [hunk 14]
```diff
@@ -56,55 +68,73 @@ auto Workspaces::update() -> void {
...
+      if (bar_.output->name == new_monitor) {  // TODO: implement this better
```
Comment (Anakael) [inline src/modules/hyprland/workspaces.cpp:109]: You can do the same trick with indention here.

File: src/modules/hyprland/workspaces.cpp [hunk 15]
```diff
@@ -136,13 +163,27 @@ Workspaces::~Workspaces() {
...
+  is_special_ = false;
+
+  if (name_.find("name:") == 0) {
+    name_ = name_.substr(5);
+  } else if (name_.find("special") == 0) {
+    name_ = id_ == -99 ? name_ : name_.substr(13);
+    is_special_ = 1;
+  }
+  // spdlog::info("\nNew workspace:\n\tid:({})\n\tname:({})\n\tmonitor:({})\n\tspecial:({})\n", id_,
+  // name_, monitor_, is_special_ ? "yes" : "no"); //make less noisy
```
Comment (Anakael) [inline src/modules/hyprland/workspaces.cpp:181]: I think it should be eather removed or changed to debug.

File: src/modules/hyprland/workspaces.cpp [hunk 16]
```diff
@@ -161,14 +202,29 @@ void Workspace::update(const std::string &format, const std::string &icon) {
...
+                if ((a->is_special()) ^ (a->is_special())) {
+                  return a->id() > b->id();
+                } else {
+                  return a->id() < b->id();
+                }
+              }
+              if ((a->id() > 0) ^ (b->id() > 0)) {
+                return a->id() > b->id();
+              }
+              spdlog::error("huh!!?");
```
Comment (Anakael) [inline src/modules/hyprland/workspaces.cpp:226]: ?))

**Review by Anakael — CHANGES_REQUESTED** [review]

File: include/modules/hyprland/workspaces.hpp (hunk 6 above)
Comment (MightyPlaza) [reply include/modules/hyprland/workspaces.hpp:50]: it is in workspaces

**Comment by Anakael** [comment]
Can you please explain what named, special and named special workspaces are?

File: include/modules/hyprland/workspaces.hpp (hunk 7 above)
Comment (MightyPlaza) [reply include/modules/hyprland/workspaces.hpp:66]: what you were getting from ipc was not an int it was a string, which for the normal workspaces is "id", but for named or special it won't work

Comment (MightyPlaza) [reply include/modules/hyprland/workspaces.hpp:66]: trying to use the use the name and get the id to the work with it would probably be worse

File: include/modules/hyprland/workspaces.hpp (hunk 8 above)
Comment (MightyPlaza) [reply include/modules/hyprland/workspaces.hpp:58]: same as above (IPC)

File: include/modules/hyprland/workspaces.hpp (hunk 6 above)
Comment (Anakael) [reply include/modules/hyprland/workspaces.hpp:50]: Oh, right. Sorry.

File: src/modules/hyprland/workspaces.cpp (hunk 10 above)
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp:72]: same (IPC gives name not id)

File: include/modules/hyprland/workspaces.hpp (hunk 7 above)
Comment (Anakael) [reply include/modules/hyprland/workspaces.hpp:66]: I see.

File: include/modules/hyprland/workspaces.hpp (hunk 12 above)
Comment (MightyPlaza) [reply include/modules/hyprland/workspaces.hpp:32]: sure, was using hyprland's name

File: src/modules/hyprland/workspaces.cpp (hunk 16 above)
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp:226]: should never get there
will probably make it unreachable

**Comment by Anakael** [comment]
> Can you please explain what named, special and named special workspaces are?

I've found it in wiki.

**Comment by MightyPlaza** [comment]
there seems to be an issue with dispatcher that makes giving no args use the dispatcher as arg, that's why clicking the non named special workspace doesn't work.
currently trying to fix it and other issues on hyprland's side

File: src/modules/hyprland/workspaces.cpp (hunk 15 above)
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp:181]: used it for debugging at start, probably will remove

File: man/waybar-hyprland-workspaces.5.scd (hunk 4 above)
Comment (MightyPlaza) [reply man/waybar-hyprland-workspaces.5.scd:64]: done

File: src/modules/hyprland/workspaces.cpp (hunk 14 above)
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp:109]: I need the 2nd if for the else

File: src/modules/hyprland/workspaces.cpp (hunk 15 above)
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp]: removed

File: src/modules/hyprland/workspaces.cpp (hunk 13 above)
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp]: fixed

File: include/modules/hyprland/workspaces.hpp (hunk 1 above)
Comment (MightyPlaza) [reply include/modules/hyprland/workspaces.hpp]: done

File: src/modules/hyprland/workspaces.cpp (hunk 2 above)
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp]: fixed

**Comment by zjeffer** [comment]
This might not be a big problem but I notice the workspace creation is much slower than the wlr/workspaces implementation. Here's a video showing my bar with wlr/workspaces at the top, and your implementation on the bottom:

https://github.com/Alexays/Waybar/assets/4633209/ace456bc-e77d-406c-95e4-b270582538f2

I quickly scroll through the workspaces, and have to wait every once in a while to let the bottom bar update. 

I built the latest master and noticed it doesn't have this problem, although it has another problem where it sometimes doesn't destroy the workspaces, resulting in something like this:

**Comment by MightyPlaza** [comment]
tested spamming the mouse wheel and it seems to update at the same speed as wlr. Also seems to be updating at the same speed as the IPC socket.
the workspaces not getting deleted should be fixed soon, despite not being able to repro

**Comment by MightyPlaza** [comment]
fixed one of the issues for clicking the main special workspace on hyprland's side.
should be in the main branch soon
Edit: also had to change previous waybar code for the intended implementation

File: src/modules/hyprland/workspaces.cpp (hunk 14 above)
Comment (Anakael) [reply src/modules/hyprland/workspaces.cpp]:
```c++
if (bar_.output->name == new_monitor) {
  workspaces_to_remove_.push_back(workspace);
  continue;
}

// rest code
```

File: src/modules/hyprland/workspaces.cpp [hunk 17]
```diff
@@ -43,68 +53,92 @@ Workspaces::Workspaces(const std::string &id, const Bar &bar, const Json::Value
...
+
   } else if (eventName == "createworkspace") {
-    int new_workspace_id;
-    std::from_chars(payload.data(), payload.data() + payload.size(), new_workspace_id);
-    workspaces_to_create_.push_back(new_workspace_id);
+    const Json::Value workspaces_json = gIPC->getSocket1JsonReply("workspaces");
+    for (Json::Value workspace_json : workspaces_json) {
+      if (workspace_json["name"].asString() == payload &&
+          (all_outputs() || bar_.output->name == workspace_json["monitor"].asString()) &&
+          (workspace_json["name"].asString().find("special:") != 0 || show_special())) {
```
Comment (Anakael) [inline src/modules/hyprland/workspaces.cpp:100]: Let's put `show_special` check at first in order to skip comparing strings if flag is set to true.

Comment (Anakael) [reply src/modules/hyprland/workspaces.cpp:100]: And on 155 line too.

File: src/modules/hyprland/workspaces.cpp [hunk 18]
```diff
@@ -136,13 +167,25 @@ Workspaces::~Workspaces() {
...
+      output_(value["monitor"].asString()),  // TODO:allow using monitor desc
+      windows_(value["id"].asInt()) {
+  active_ = true;
+  is_special_ = false;
+
+  if (name_.find("name:") == 0) {
+    name_ = name_.substr(5);
+  } else if (name_.find("special") == 0) {
+    name_ = id_ == -99 ? name_ : name_.substr(13);
+    is_special_ = 1;
```
Comment (Anakael) [inline src/modules/hyprland/workspaces.cpp:182]: `is_special = true`? )

**Review by Anakael — CHANGES_REQUESTED** [review]

File: src/modules/hyprland/workspaces.cpp (hunk 18 above)
Comment (zjeffer) [reply src/modules/hyprland/workspaces.cpp:182]: I changed this here: https://github.com/MightyPlaza/Waybar/pull/3

File: src/modules/hyprland/workspaces.cpp (hunk 14 above)
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp]: not inside a loop, can't use continue

File: src/modules/hyprland/workspaces.cpp (hunk 16 above)
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp]: removed it

File: src/modules/hyprland/workspaces.cpp (hunk 17 above)
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp]: done

**Comment by MightyPlaza** [comment]
I suggest this gets merged now, since the only issue is the non removed workspaces which only happen on spamming the mouse wheel.
That should be fixed, in my opinion, by using a vector of combined DTO to create/destroy, which will likely be a major rewrite, so will take some time.
What do you guys think?

**Comment by zjeffer** [comment]
> since the only issue is the non removed workspaces which only happen on spamming the mouse wheel.

To clarify: this problem only occured for me on the master branch, not on this PR's branch. On this PR's branch, it's just a little slower when creating & destroying hundreds of workspaces per second. Not a big deal.

I agree it can be merged now.

File: include/modules/hyprland/workspaces.hpp (hunk 9 above)
Comment (MightyPlaza) [reply include/modules/hyprland/workspaces.hpp]: this and a combined destroy/remove queue is something that should be done, but maybe in the future, since it'll require a rewrite of most of the code.

**Review by Anakael — APPROVED** [review]

**Review by zjeffer — APPROVED** [review]

**Comment by zjeffer** [comment]
@Alexays  can this be merged? I added persistent workspaces on my own fork and would like to open a separate PR for this.

**Comment by Alexays** [comment]
Yes, LGTM

**Comment by sudg6** [comment]
on-click: activate is not working, nor can I click to switch to any workspace button!

[Edit]
`pacman` manager is still hosting an outdated version of Waybar.
I compiled it under my Arch Linux machine and had an issue with loading the libcanva.so library but manually fixed it as discussed [here](https://github.com/Alexays/Waybar/issues/2281#issuecomment-4818689311).

**Comment by zjeffer** [comment]
> on-click: activate is not working, nor can click to switch to any workspace button!

This has been fixed in https://github.com/Alexays/Waybar/pull/5013, @Alexays just needs to make a new release, or you can compile it yourself.

**Comment by zjeffer** [comment]
@sudg6 your edit is incorrect 

>  pacman manager is still hosting an outdated version of Waybar.

this is not Arch's or pacman's fault. There simply isn't a release yet that includes the fix.
